### ISSUE TITLE
Function with generic functional argument specialized with literals

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#4985]
### Discussed in https://github.com/microsoft/pyright/discussions/4984

<div type='discussions-op-text'>

<sup>Originally posted by **sylee957** April 20, 2023</sup>
Given the generic object `Tree` and the constructor function `tree`

```python
from dataclasses import dataclass
from typing import Callable, Generic, Literal, TypeVar

S = TypeVar("S")
T = TypeVar("T")


@dataclass(frozen=True)
class Tree(Generic[S, T]):
    left: S
    right: T


def tree(s: S, t: T) -> Tree[S, T]:
    return Tree(s, t)
```

`test_1` does not passes the type check with the error `"int" cannot be assigned to type "Literal[2]"`

```python
def test_1(t: T, f: Callable[[Tree[T, Literal[2]]], None]) -> None:
    return f(tree(t, 2))


def test_2(t: T, f: Callable[[Tree[T, Literal[2]]], None]) -> None:
    return f(Tree(t, 2))
```

I'm using pyright 1.1.304</div>

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.306, which I just published. It will also be included in a future release of pylance.
